### PR TITLE
Initial code commit

### PR DESCRIPTION
**Pull request by richerve**
Firs commit with all changes required, in a working state and ready for review.

Will include sample k8s manifests in our private repo to test this in Kanopy.

### Design choices on Grafana client

I decided to wrap the [gapi](github.com/grafana/grafana-api-golang-client) client library as in the future it would be easy to refactor if that library is decommissioned, and also to add better handling of certain cases like returning `nil` instead of empty struct or adding new methods (UpsertOrgUser).

Unit tests on Grafana client are somewhat basic, using a mock, as the upstream library (gapi) already tests actual http calls and responses.

### CODE REVIEW

File: pkg/grafana/client.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			Password: "abc123",
+		})
+		if err != nil {
+			log.Error(err)
+			return err
+		}
+		log.Infof("new user %s created with id %d", user.Login, uid)
+	}
+
+	err = c.AddOrgUser(OrgId, user.Login, role)
```
Comment (dskatz) [inline pkg/grafana/client.go:92]: How does this handle the ` grafana_admin = true`  from our configuration?

**Review by dskatz — COMMENTED** [review]
Some comments about structure.

There are some other items we need to consider. https://github.com/mongodb-forks/grafana/commit/72adda95ccb5c5dcb5770eed6732c5dec7d2436a

As referenced in the commit above, we need to consider if the role is assignable to the user based on a hierarchy.  This keeps higher level roles assigned to a user since a user can be in multiple groups  that are given either Admin, Editor, and Viewer.  If the viewer role is the last processed the user essentially has their access revoked.

File: internal/cli/cli.go [hunk 2]
```diff
@@ -23,11 +28,31 @@ func NewRootCommand() *cobra.Command {
...
 	return cmd
 }
 
 func (c *RootCommand) persistentPreRunE(cmd *cobra.Command, args []string) error {
+
+	// Read from config
+	viper.SetConfigName("config")
+	viper.AddConfigPath(".")
+	viper.AddConfigPath("/etc/grafana-auth-proxy/")
+	viper.WatchConfig()
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/cli/cli.go:47]: I haven't fully tested this but the feature I'm looking to have is us modifying the configmap to have a new group mapping, the file gets updated in the pod filesystem and viper re-reads the config without needing to restart the proxy. I can add a function hook to show a message in logs as is described [here](https://github.com/spf13/viper#watching-and-re-reading-config-files).

The groups config is only read when the proxy is activated, basically when a request comes, so config should be as updated as possible to avoid a case where the user waits because a config hasn't been applied.

File: internal/config/config.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+type Groups map[string]Group
+
+type Org struct {
+	OrgId int64  `json:"orgId"`
+	Role  string `json:"role"`
+}
+
+func ParseGroups() (Groups, error) {
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/config/config.go:16]: do you mean that I should simply call `viper.UnmarshalKey` when required, like in the proxy call?

File: internal/server/server.go [hunk 4]
```diff
@@ -1,24 +1,110 @@
...
+		// mapping in configuration
+		validUserGroups := config.UserGroupsInConfig(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", claims.Subject, validUserGroups)
+
+		if len(validUserGroups) > 0 {
+			for _, group := range validUserGroups {
+				data := s.groups[group]
+
+				for _, org := range data.Orgs {
+					orgUser := gapi.User{
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:76]: good catch, thanks

File: pkg/grafana/client.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+			return err
+		}
+	}
+
+	if foundUser == nil {
+		log.Infof("no user with login %s found, creating new one", user.Login)
+		uid, err := c.CreateUser(gapi.User{
+			Login:    user.Login,
+			Email:    user.Email,
+			Password: "abc123",
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:83]: It's here to satisfy the API. I'll generate a secure password but it would be essentially the same for all new users.

As how the password can be used, we are essentially going to disable the local login mechanisms of Grafana as we serve it to end users, using Ingress I mean. I can test if I can use those passwords when bypassing Ingress, basically not using the auth proxy.

File: pkg/grafana/client.go (hunk 1 above)
Comment (richerve) [reply pkg/grafana/client.go:92]: good point, currently it doesn't handle that case. That flag is part of the `gapi.OrgUser` object, that I'm not using directly to try to avoid having more than one definition of a user. Will figure out how to add support for that.

File: pkg/grafana/client.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		if err != nil {
+			log.Error(err)
+			return err
+		}
+		log.Infof("new user %s created with id %d", user.Login, uid)
+	}
+
+	err = c.AddOrgUser(OrgId, user.Login, role)
+	if err != nil {
+		if !strings.Contains(err.Error(), "User is already member of this organization") {
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:94]: If a user is already part of the organization then upsert must return no error. Here I'm taking into account errors that are different than the one returned when a user is already a member. As I understand those later errors will be failures when calling the Grafana API.

[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/grafana/client.go:94]: I think I read this wrong. Its "not" user already a member.

File: pkg/grafana/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/grafana/client.go:83]: Then its probably best practice to to generate a secure random string and leave it at that.

File: pkg/grafana/client.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		if !strings.Contains(err.Error(), "User not found") {
+			return err
+		}
+	}
+
+	if foundUser == nil {
+		log.Infof("no user with login %s found, creating new one", user.Login)
+		uid, err := c.CreateUser(gapi.User{
+			Login:    user.Login,
```
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/grafana/client.go:81]: Ok, at least I was consistent and read these conditionals wrong as well.  "not" contains user not found.

File: internal/config/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/config/config.go:16]: Right now, ParseGroups is only called once in the runE.

File: internal/cli/cli.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/cli/cli.go:47]: To accomplish that you will need to look at using a sync.Map and how group and config information gets passed to the server also may need adjustment to prevent race conditions.

File: internal/config/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/config/config.go:16]: removed

File: internal/server/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:76]: done

File: pkg/grafana/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:81]: I can change the logic to start with `foundUser != nil`

File: pkg/grafana/client.go (hunk 1 above)
Comment (richerve) [reply pkg/grafana/client.go:102]: added support for grafana_admin

File: pkg/grafana/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go]: done

File: pkg/grafana/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go]: I updated the logic on this, more info in https://github.com/kanopy-platform/grafana-auth-proxy/pull/1/commits/ac5468df479d847539714842735bdec4248bb836

File: internal/config/config.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+type Group struct {
+	Orgs []Org `json:"orgs"`
+}
+
+type Groups map[string]Group
+
+type Org struct {
+	ID           int64  `json:"id"`
+	Role         string `json:"role"`
+	GrafanaAdmin *bool  `json:"grafanaAdmin"`
```
Comment (whwalter) [inline internal/config/config.go:12]: Nit: Why does this need to be a pointer?

File: internal/server/server.go [hunk 9]
```diff
@@ -1,24 +1,152 @@
...
+
+			// Update ID on orgUser
+			orgUser.ID = uid
+		} else {
+			orgUser.ID = foundUser.ID
+		}
+
+		if len(validUserGroups) > 0 {
+			for _, group := range validUserGroups {
+				if groupConfig, ok := s.groups[group]; ok {
```
Comment (whwalter) [inline internal/server/server.go:106]: If UserGroupsInConfig returned Groups then you could remove the if s.groups[group] check.

File: internal/server/server.go [hunk 10]
```diff
@@ -1,24 +1,152 @@
...
+					for _, org := range groupConfig.Orgs {
+						// Check if the users has a more permissive role and apply that instead
+						if !grafana.IsRoleAssignable(userOrgsRole[org.ID], models.RoleType(org.Role)) {
+							continue
+						}
+
+						userOrgsRole[org.ID] = models.RoleType(org.Role)
+
+						if org.GrafanaAdmin != nil && *org.GrafanaAdmin {
```
Comment (whwalter) [inline internal/server/server.go:115]: Nit: Is there a reason GrafanaAdmin needs to be a *bool?

File: pkg/grafana/client.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	UserByEmail(email string) (user gapi.User, err error)
+	CreateUser(user gapi.User) (int64, error)
+	AddOrgUser(orgID int64, user, role string) error
+	UpdateOrgUser(orgID, userID int64, role string) error
+}
+
+func NewClient(baseURL *url.URL, cfg gapi.Config) (*Client, error) {
+	newClient := &Client{}
+
+	client, err := gapi.New(baseURL.String(), cfg)
```
Comment (whwalter) [inline pkg/grafana/client.go:32]: You probably want to set https://github.com/grafana/grafana-api-golang-client/blob/9bd3cddb7c2b00364d2c55e39c3db7640558c6b7/client.go#L23 to a reasonably configured *http.Client with an explicit timeout so that we pass an error from the proxy back to the client and not a 503 from the load balancer.

File: internal/server/server.go [hunk 12]
```diff
@@ -1,24 +1,139 @@
...
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/config"
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/jwt"
+	"github.com/kanopy-platform/grafana-auth-proxy/pkg/grafana"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	grafanaAuthHeader = "X-WEBAUTH-USER"
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:18]: good point, will make it configurable

File: internal/server/server.go [hunk 13]
```diff
@@ -1,24 +1,139 @@
...
+		}
+
+		// Get claims from token
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		log.Infof("attempting to log in user %s", claims.Subject)
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:64]: I'll include more checks

File: internal/server/server.go [hunk 14]
```diff
@@ -1,24 +1,152 @@
...
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error looking for user")
+			return
+		}
+
+		if foundUser == nil {
+			log.Infof("no user with login %s found, creating new one", orgUser.Login)
+			uid, err := s.grafanaClient.CreateUser(orgUser)
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error creating new user")
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:94]: good catch, thx

File: internal/server/server.go [hunk 15]
```diff
@@ -1,24 +1,139 @@
...
+				return
+			}
+
+			// Update ID on orgUser
+			orgUser.ID = uid
+		} else {
+			orgUser.ID = foundUser.ID
+		}
+
+		if len(validUserGroups) > 0 {
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:104]: Agreed, it should be unauthorized if the user is not on a group associated to an org in Grafana

File: internal/server/server.go (hunk 10 above)
Comment (richerve) [reply internal/server/server.go:115]: to avoid needing to pass it in config everytime, no?

File: internal/server/server.go [hunk 16]
```diff
@@ -1,24 +1,152 @@
...
+		log.Infof("attempting to log in user %s", claims.Subject)
+		log.Debugf("claim groups for user %s: %v", claims.Subject, claims.Groups)
+
+		// validUserGroups represents the intersection of user groups from claim with the group
+		// mapping in configuration
+		validUserGroups := config.UserGroupsInConfig(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", claims.Subject, validUserGroups)
+
+		// Creat gapi.User for lookup purposes
+		orgUser := gapi.User{
```
Comment (dskatz) [inline internal/server/server.go:73]: Why do we create a new `orgUser` here instead of just checking `foundUser`?

File: internal/server/server.go [hunk 17]
```diff
@@ -1,24 +1,152 @@
...
+			uid, err := s.grafanaClient.CreateUser(orgUser)
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error creating new user")
+				return
+			}
+
+			// Update ID on orgUser
+			orgUser.ID = uid
+		} else {
+			orgUser.ID = foundUser.ID
```
Comment (dskatz) [inline internal/server/server.go:101]: Why not just use foundUser?

File: pkg/grafana/client.go [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+// UpsertOrgUser adds a user to an Organization if not present or
+// updates the user role if already a member.
+func (c *Client) UpsertOrgUser(orgID int64, user gapi.User, role string) error {
+
+	var isOrgMember bool
+
+	err := c.AddOrgUser(orgID, user.Login, role)
+	if err != nil {
+		if !strings.Contains(err.Error(), "User is already member of this organization") {
+			log.Error(err)
```
Comment (dskatz) [inline pkg/grafana/client.go:103]: Remove this log entry.

File: pkg/grafana/client.go [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+			return err
+		} else {
+			isOrgMember = true
+		}
+	}
+
+	if isOrgMember {
+		// Update user if it's a member in case roles changed in config
+		log.Infof("updating user, %s", user.Login)
```
Comment (dskatz) [inline pkg/grafana/client.go:112]: Remove this log entry.

**Review by dskatz — COMMENTED** [review]
I think the gapi.User should just be a non-pointer everywhere.  This makes the client API more consistent and removes the need to create an `orgUser` that copies the UID from the foundUser.

There also needs to be more test coverage. I think we are missing key scenarios around org to group mapping and role assignments. 

INFO and ERROR logging I think should also be consolidated to the `server` and not in the client.

File: internal/cli/cli.go [hunk 20]
```diff
@@ -23,11 +28,31 @@ func NewRootCommand() *cobra.Command {
 
 	cmd.PersistentFlags().String("log-level", "info", "Configure log level")
 	cmd.PersistentFlags().String("listen-address", ":8080", "Server listen address")
+	cmd.PersistentFlags().Bool("tls-skip-verify", false, "Skip TLS certificate verification")
+	cmd.PersistentFlags().String("grafana-proxy-url", "", "Grafana url to proxy to")
+	cmd.PersistentFlags().String("grafana-client-url", "", "Grafana url to connect with client")
+	cmd.PersistentFlags().String("cookie-name", "", "Cookie name with jwt token. If set will take precedence over auth header")
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/cli/cli.go:34]: Correct, I'll set `auth_token` as default value. I could also let this empty and panic if a value is not passed. What do you think?

File: pkg/grafana/client.go [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+	client GAPIClient
+}
+
+type GAPIClient interface {
+	UserByEmail(email string) (user gapi.User, err error)
+	CreateUser(user gapi.User) (int64, error)
+	AddOrgUser(orgID int64, user, role string) error
+}
+
+func NewClient(baseURL *url.URL, cfg gapi.Config) (*Client, error) {
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:29]: what David says was my thinking when writing this

File: internal/config/config.go (hunk 8 above)
Comment (richerve) [reply internal/config/config.go:12]: answered on the same question done when the implementation happens

File: internal/server/server.go (hunk 16 above)
Comment (richerve) [reply internal/server/server.go:73]: true, it can simply use foundUser

File: internal/server/server.go [hunk 22]
```diff
@@ -1,24 +1,152 @@
...
+
+		// Creat gapi.User for lookup purposes
+		orgUser := gapi.User{
+			Login: claims.Subject,
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
+
+		if claims.Email != "" {
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:80]: isn't the Email claim optional?

File: pkg/grafana/client.go [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+	// gapi returns an empty struct
+	if user.Login == "" {
+		return nil, nil
+	}
+
+	return &user, nil
+}
+
+// CreateUser adds a new global user to Grafana
+func (c *Client) CreateUser(user gapi.User) (int64, error) {
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:60]: The reason I did that was to have LookupUser return a nil if a user is not present, but it can be an empty user so I agreed that working with objects instead of references everywhere makes more sense

File: internal/cli/cli.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/cli/cli.go:35]: Sane defaults are good. 

e.g. grafana-proxy-url. `http://grafana.example.com`

File: internal/cli/cli.go [hunk 24]
```diff
@@ -48,10 +74,75 @@ func (c *RootCommand) persistentPreRunE(cmd *cobra.Command, args []string) error
...
+
 func (c *RootCommand) runE(cmd *cobra.Command, args []string) error {
+
 	addr := viper.GetString("listen-address")
+	log.Infof("listening on %s", addr)
+
+	opts := defaultServerOptions()
+
+	adminUser := viper.GetString("admin-user")
+	adminPassword := viper.GetString("admin-password")
```
Comment (dskatz) [inline internal/cli/cli.go:100]: nit: this can probably be combined  / moved to where BasicAuth is configured in the gapi.Config object.

File: internal/cli/cli.go [hunk 25]
```diff
@@ -48,10 +74,75 @@ func (c *RootCommand) persistentPreRunE(cmd *cobra.Command, args []string) error
...
+
+	grafanaProxyUrlRaw := viper.GetString("grafana-proxy-url")
+	grafanaProxyURL, err := url.Parse(grafanaProxyUrlRaw)
```
Comment (dskatz) [inline internal/cli/cli.go:103]: nit:  grafanaProxyUrlRaw is not needed.

File: internal/cli/cli.go [hunk 26]
```diff
@@ -48,10 +74,75 @@ func (c *RootCommand) persistentPreRunE(cmd *cobra.Command, args []string) error
...
+	if err != nil {
+		log.Error("grafana-proxy-url is not a proper url")
+		return err
+	}
+	opts = append(opts, server.WithGrafanaProxyURL(grafanaProxyURL))
+
+	grafanaClientUrlRaw := viper.GetString("grafana-client-url")
```
Comment (dskatz) [inline internal/cli/cli.go:110]: nit: grafanaClientUrlRaw is not needed.

File: pkg/grafana/client.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/grafana/client.go:59]: I agree, but after seeing it in use and the disconnect it creates with the other functions, it loses that nil check benefit, since you already need to test for an "empty" object in other areas.

File: internal/config/config.go (hunk 8 above)
Comment (dskatz) [reply internal/config/config.go:12]: I am not entirely clear on the answer here either.  It was a bool pointer in the grafana fork, because that is how it is used in their internal code base.  I don't see what a nil check gets us here. If not present in configuration, it defaults to false.

File: pkg/grafana/client.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:59]: agreed, updated all methods to conform to the same api

File: pkg/grafana/client.go [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+type GAPIClient interface {
+	UserByEmail(email string) (user gapi.User, err error)
+	CreateUser(user gapi.User) (int64, error)
+	AddOrgUser(orgID int64, user, role string) error
+}
+
+func NewClient(baseURL *url.URL, cfg gapi.Config) (*Client, error) {
+	newClient := &Client{}
+
+	client, err := gapi.New(baseURL.String(), cfg)
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client.go:36]: done

File: internal/cli/cli.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (whwalter) [reply internal/cli/cli.go]: Sane defaults for everything. If a setting is required and doesn't have a sane default, fail with an error but don't intentionally panic.

File: internal/config/config.go (hunk 8 above)
Comment (whwalter) [reply internal/config/config.go:12]: This would not default to false from configuration it would default to nil as a pointer default value. If it were a bool it would default to false.

Comment (dskatz) [reply internal/config/config.go:12]: Right, that's my point.  Why use a pointer bool, the nil check in this case seems redundant.

File: internal/server/server.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:107]: fixed

File: internal/server/server.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/server/server.go]: Yes, the email claim is optional as per the OIDC spec.  The default value of the orgUser.Email is also an empty string.  So is their value in checking if the `claims.Email` contains a value? Since if not included it would also be an empty string.

If both are empty, isn't this `orgUser.Email = claims.Email` always valid?

[the remark this answers is not in the document]
Comment (whwalter) [reply internal/server/server.go]: Do we care about pushing the email claim to the grafana global and org users? It doesn't seem like we do since I imagine it is used for maybe password reset which is hopefully disabled when this proxy is enabled.

[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: correct, it's not a hard requirement but I thought it would be useful if the information is available on the claims.

[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: applied David's suggestion

File: internal/server/server.go (hunk 17 above)
Comment (richerve) [reply internal/server/server.go]: applied, used orgUser to build the user object

File: internal/server/server.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/server/server.go]: I think if the auth proxy takes full control over how a user authenticates and user info, it should make sure the grafana user object accurately reflects the info in the jwt.

[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: agreed, it will basically populate as much as it can from the auth source, in this case the jwt

File: internal/server/server.go (hunk 9 above)
Comment (richerve) [reply internal/server/server.go]: applied your suggestion

File: internal/config/config.go [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// UserGroupsInConfig matches the user groups (from claims) that are
+// present in config
+func UserGroupsInConfig(userGroups []string, groups Groups) []string {
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/config/config.go]: applied your suggestion

File: internal/server/server.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: included explicit checks for Subject and Groups and return Unauthorized in case those are missing

File: internal/server/server.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (whwalter) [reply internal/server/server.go]: We only upsert org users so if that is the case should this be an update on change to existing global users also?

File: internal/cli/cli.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/cli/cli.go]: applied defaults for all I think was sensible, for example admin-password doesn't have default and the program now fails if not present

File: internal/server/server.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: every user is an Org User by definition. Users are at least assigned to the default Org (1). In the code logic the user is always updated on the corresponding Org, even if that's Org 1, so the update path will always trigger and affect what is considered a "global user".

File: pkg/grafana/client.go (hunk 11 above)
Comment (richerve) [reply pkg/grafana/client.go:36]: - What would be sensible values? I noticed that in gapi they use https://github.com/hashicorp/go-cleanhttp/blob/master/cleanhttp.go, that it seems to contain those values.
- Should NewClient provide those default values or set default on the cli and make some configurable with extra parameters?

Comment (whwalter) [reply pkg/grafana/client.go:36]: Huh, I missed that they are already setting defaults in their client constructor. These might be fine I just wanted some timeout vs wait forever.

Comment (whwalter) [reply pkg/grafana/client.go:36]: If you wanted to tune them, your proxy has < 60 seconds to complete the entire login handshake and have grafana respond with a session. These defaults are too long for the timeout response to be returned before the ELB hangs up so tuning them to be short enough that the entire flow, including the response from the final request to grafana, is ensured to fit into 60s would be the desired defaults. That might be a premature optimization though.

File: internal/server/server.go (hunk 10 above)
Comment (richerve) [reply internal/server/server.go]: removed the pointer and used a simple bool instead

File: internal/server/server.go [hunk 29]
```diff
@@ -1,24 +1,152 @@
...
+		validUserGroups := config.UserGroupsInConfig(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", claims.Subject, validUserGroups)
+
+		// Creat gapi.User for lookup purposes
+		orgUser := gapi.User{
+			Login: claims.Subject,
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: moved

File: internal/server/server.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go]: done

**Comment by richerve** [comment]
@whwalter added timeout as a configurable parameters with default value of 60s

File: internal/server/server.go [hunk 30]
```diff
@@ -1,24 +1,168 @@
...
+			}
+
+			orgUser.ID = uid
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
+
+		for _, groupConfig := range validUserGroups {
+			if groupConfig.GrafanaAdmin && !orgUser.IsAdmin {
```
Comment (dskatz) [inline internal/server/server.go:118]: This seem like we will not be able to revoke GrafanaAdmin in configuration.

**Review by dskatz — COMMENTED** [review]
A couple of minor comments.  I think we need to expand test coverage across the server and grafana client.  

It may also be good to include a regression test that has more org and group configurations. E.g. if a user is in more than one group with different permissions in multiple orgs, e2e do they get have the correct role in each org.

**Comment by richerve** [comment]
that's correct on the GrafanaAdmin revocation aspect

File: internal/server/server.go [hunk 31]
```diff
@@ -1,24 +1,168 @@
...
+				userOrgsRole[org.ID] = models.RoleType(org.Role)
+			}
+		}
+
+		for orgID, role := range userOrgsRole {
+			err = s.grafanaClient.UpsertOrgUser(orgID, orgUser, string(role))
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error upserting user")
+				return
+			}
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:142]: I understand what you're saying. Here are the 2 options I see:

- Have a default policy of allowing the user even if cannot be upserted and use the default values in Grafana for auto org number and auto org assign role.
- Handle default org and role logic inside the proxy, so we can compare with explicit values and let it through if they match, like default orgid 1 and default role Viewer.
- Treat the current behaviour as a known error, so users won't be authorized if the don't have a valid org assigned in our config. Basically without hidden defaults but all explicit in config. If the org is incorrect then we would need to fix it, same reason as why there's no automatic org management handling in the proxy.

File: internal/server/server.go (hunk 30 above)
Comment (richerve) [reply internal/server/server.go]: fixed

File: internal/server/server.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply internal/server/server.go:142]: I like a combination of those options.  Support a default behavior in the auth proxy config map.

[the remark this answers is not in the document]
Comment (whwalter) [reply internal/server/server.go:142]: We do a GetUser -> CreateUser flow to create the 'global' grafana user on lines 91-112 so if we make it here we have a global grafana user alreadt. I assume this user is by default viewer in the default org but that should be confirmed. This produces the minimum needed for minting the session token and grafana authentication so it should be safe to log these errors and continue.

If a group -> role mapping fails here it will result in that user either not having access to an org or having less permission than they should. Either way, they should still be able to authenticate to grafana with missing permissions vs getting an 'unauthorized' error.

[the remark this answers is not in the document]
Comment (richerve) [reply internal/server/server.go:142]: for point 1, the assumption of default viewer in default org is incorrect, I mean, as I mentioned on my 1st and 2nd points that can either happen after Grafana has "realized" the new user, that meaning that the proxy won't know about those default values and will simply allow the user to passthrough and let Grafana handle defaults. Or the 2nd option is to move that logic to the proxy so we say what org and role are defaults and explicitly assign it, if there's no other Org that matches or as default value for all.

[the remark this answers is not in the document]
Comment (whwalter) [reply internal/server/server.go:142]: Screen Shot 2021-11-23 at 3 17 33 PM

If there are 0 successful groups upserts and we still let the user in it would have the above permission.

File: pkg/grafana/client_test.go [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+		ID:    1,
+		Email: "foo@example.com",
+		Login: "foo",
+	}
+}
+
+func TestLookupUser(t *testing.T) {
+	user := setupUser()
+
+	client := NewMockClient(user, nil)
```
[the remark this answers is not in the document]
Comment (richerve) [reply pkg/grafana/client_test.go:23]: added

File: internal/server/server.go [hunk 33]
```diff
@@ -1,24 +1,123 @@
...
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		if claims.Subject == "" {
+			logAndError(w, http.StatusUnauthorized, err, "sub claim is empty")
+			return
+		}
```
Comment (whwalter) [inline internal/server/server.go:66]: These ifs are uncovered in tests and the logic that follows depends on a number of assumptions about the value of claims.Subject so we should add a test to assert these fail properly.

- No cookie
- invalid jwt
- jwt with an empty Subject claim

Comment (richerve) [reply internal/server/server.go:66]: makes sense

File: pkg/grafana/client_test.go [hunk 34]
```diff
@@ -0,0 +1,162 @@
...
+	// test adding to new org
+	err := client.AddOrgUser(2, "foo", "Editor")
+	assert.NoError(t, err)
+
+	// test already a member
+	err = client.AddOrgUser(1, "foo", "Editor")
+	assert.Contains(t, err.Error(), "User is already member")
+}
+
+func TestUpsertOrgUser(t *testing.T) {
```
Comment (dskatz) [inline pkg/grafana/client_test.go:61]: This is missing two test cases:
 * grafanaAdmin
 * IsRoleAssignable 
 
 
 With the addition of these, it may be worth updating this to a table test format.

Comment (richerve) [reply pkg/grafana/client_test.go:61]: agreed, I think other tests will also be better using table testing

**Comment by richerve** [comment]
@whwalter I improved the coverage in server and added the test cases you mentioned

**Comment by richerve** [comment]
@dskatz, IsRoleAssignable is both unit tested for all cases and in `TestUpdateOrgUserAuthz` I test that the most permissive role is actually assigned, so I think that's covered.

I added an explicit test for grafanaAdmin in the aformentioned test function that uses that.

File: internal/config/config_test.go [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+					ID:   1,
+					Role: "Editor",
+				},
+			},
+		},
+	}
+
+	validGroups := ValidUserGroups(userGroups, groups)
+	assert.Contains(t, validGroups, "two")
+	assert.NotContains(t, validGroups, "one")
```
[the remark this answers is not in the document]
Comment (richerve) [reply internal/config/config_test.go]: fixed

File: pkg/grafana/client_test.go [hunk 36]
```diff
@@ -0,0 +1,209 @@
...
+	// the client is only used to update grafana admin permissions in this case
+	// so it doesn't matter what's the current value of user or orgMap is
+	client := NewMockClient(user, userOrgsRoleMap{})
+
+	tests := []struct {
+		groups   config.Groups
+		expected userOrgsRoleMap
+		fail     bool
+	}{
+		{
```
Comment (dskatz) [inline pkg/grafana/client_test.go:117]: One additional test to evaluate this critical path is needed:
```
			if !isRoleAssignable(userOrgsRole[org.ID], models.RoleType(org.Role)) {
				continue
			}
```

e.g.
```
		{
			groups: config.Groups{
				"foo": {
					Orgs: []config.Org{
						{
							ID:   1,
							Role: "Admin",
						},
					},
				},
				"bar": {
					Orgs: []config.Org{
						{
							ID:   1,
							Role: "Editor",
						},
					},
				},
			},
			expected: userOrgsRoleMap{1: "Admin"},
		},
```

**Review by dskatz — APPROVED** [review]
LGTM

**Review by whwalter — APPROVED** [review]
